util/kernel_names.h

```diff
@@ -0,0 +1,139 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Provide adapter for safely using different types with kernel names in
+//  generic code
+//
+*******************************************************************************/
+
+#ifndef __SYCLCTS_UTIL_KERNEL_NAMES_H
+#define __SYCLCTS_UTIL_KERNEL_NAMES_H
+
+#include <string>
+#include <utility>
+
+namespace sycl_cts {
+namespace util {
+namespace kernel_name {
+
+namespace details {
+
+template <typename T>
+struct adapter_core {};
+
+struct adapter_int8_t {};
+struct adapter_int16_t {};
+struct adapter_int32_t {};
+struct adapter_int64_t {};
+struct adapter_uint8_t {};
+struct adapter_uint16_t {};
+struct adapter_uint32_t {};
+struct adapter_uint64_t {};
+
+struct adapter_byte {};
+struct adapter_size_t {};
+
+/** @brief Provides type that is safe to use in kernel names based on type given
+ *  @details There is a plenty of restrictions on types we may use for kernel
+ *           naming. The most obvious is non-forward-declarable classes, but
+ *           there are more of them.
+ *           This function uses discarded statements to map input type into the
+ *           return type we may use for the kernel name template parameters.
+ *           Note: it does not provide different kernel names for the type
+ *           aliases, so we may have the same kernel name
+ *             - for `int` and `int32_t`, or
+ *             - for `size_t` and `uint64_t`, or
+ *             - for any other possible type alias
+ *           on any specific SYCL implementation. We may say it provides adapter
+ *           aliases for the type aliases.
+ */
+template <typename T>
+constexpr auto adapter_helper() {
+  // Fixed-width types are optional and implementation-defined
```

Comment:
That is true for general C/C++ but what will happens then elsewhere in the CTS, where such types might be used without any guard?
And in the SYCL 2020 specification *4.14. Data types* & *5.5. Built-in scalar data types* it looks like we assume that they do exist, or at least the 8-bit `char` does exist, so `std::int8_t`...

Reply:
There is an unresolved issue against the SYCL 2020 spec that relates to this topic:

https://gitlab.khronos.org/sycl/Specification/-/issues/544

That issue basically says that the SYCL spec has several APIs that use the fixed-width integral types (e.g. `std::uin32_t`), but the core C++ specs says these types are optional.  We need to decide if the SYCL spec requires the underlying C++ compiler to define these types, or we need to change the SYCL spec to stop using them.

Reply:
C++20 adopts 2-complement integer representation, so I guess future version of SYCL will follow this too.

What to do if there is no `std::int16_t`, etc? But alternatively how to know the CTS works if there is no SYCL implementation for such a device without a `std::int16_t`, etc? How to test the CTS implementation itself?